### PR TITLE
fixes #14: Add schema reference into Data Events

### LINKED ISSUE #14: Add schema reference into Data Events

### PR DESCRIPTION
**Pull request by conker84**
Fixes #14 

Add schema reference into Data Events.

## Proposed Changes (Mandatory)

A brief list of proposed changes in order to fix the issue:

 - Added class `StreamsConstraintsService` that gather the schema information every `X minutes` (value configurable via streams.source.schema.polling.interval, default 300000 millis = 5 minutes)
 - Added the `schema` field into the CDC event; this provides info about constraints attached to node labels and relationships types
 - Added Jackson Kotlin module in order to simplify the integration with Jackson's APIs
 - Updated documentation
 - Remove the handcrafted mocks from `StreamsTransactionEventHandlerTest.kt` and `StreamsTransactionEventHandlerRelTest.kt` and switched to `Mockito` the only that remains is the `MockStreamsEventRouter` that provides a way to easily test some use-cases
 - Some code clean

### CODE REVIEW

File: producer/src/main/kotlin/streams/events/StreamsEventBuilder.kt [hunk 1]
```diff
@@ -167,11 +167,36 @@ class RelationshipPayloadBuilder() {
...
+        this.constraints = constraints
+        return this
+    }
+
+    private fun mapPropertiesToTypes(properties: RecordChange?): Map<String, String> {
+        return properties?.properties
+                ?.mapValues {
+                    val clazz = it.value::class
+                    if (clazz.java.isArray) {
+                        "${it.value::class.java.componentType}[]"
```
Comment (jexp) [inline producer/src/main/kotlin/streams/events/StreamsEventBuilder.kt]: missing simpleName?

File: producer/src/main/kotlin/streams/StreamsConstraintsService.kt [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+    private val job: Job
+
+    init {
+        job = GlobalScope.launch(Dispatchers.IO) {
+            while (isActive) {
+                StreamsUtils.ignoreExceptions({
+                    db.beginTx().use {
+                        db.schema().constraints
+                                .filter { try { it.label; true } catch (e: IllegalStateException) { false } }
```
[the remark this answers is not in the document]
Comment (conker84) [reply producer/src/main/kotlin/streams/StreamsConstraintsService.kt]: It happens for node constraints here:

https://github.com/neo4j/neo4j/blob/2ffd6c277db1e40c45962591b7345db49db09203/community/kernel/src/main/java/org/neo4j/kernel/impl/coreapi/schema/NodeConstraintDefinition.java#L63-L67


and for relationship constraints here:

https://github.com/neo4j/neo4j/blob/2ffd6c277db1e40c45962591b7345db49db09203/community/kernel/src/main/java/org/neo4j/kernel/impl/coreapi/schema/RelationshipConstraintDefinition.java#L39-L43

File: producer/src/main/kotlin/streams/StreamsTransactionEventHandler.kt [hunk 3]
```diff
@@ -192,9 +251,15 @@ class StreamsTransactionEventHandler(val router : StreamsEventRouter, val config
...
     override fun beforeCommit(txd: TransactionData): PreviousTransactionData {
+        val nodeConstraints = streamsConstraintsService.allForLabels()
+        val relConstraints = streamsConstraintsService.allForRelationshipType()
         var builder = PreviousTransactionDataBuilder()
+                .withNodeConstraints(nodeConstraints)
+                .withRelConstraints(relConstraints)
+
         builder = buildNodeChanges(txd, builder)
-        builder = buildRelationshipChanges(txd, builder)
+        builder = buildRelationshipChanges(txd, builder, nodeConstraints)
```
[the remark this answers is not in the document]
Comment (conker84) [reply producer/src/main/kotlin/streams/StreamsTransactionEventHandler.kt]: There is a part of the computation that is executed outside the builder (L175-L176):

https://github.com/neo4j-contrib/neo4j-streams/blob/37a80b7965c9cd6fac59a77150fba65ffe7584e5/producer/src/main/kotlin/streams/StreamsTransactionEventHandler.kt#L167-L204

Do you want to refactor this part?

**Comment by jexp** [comment]
Looks good, thanks a lot.

**Pull request merged**